packages/files/src/driver.ts

```diff
@@ -0,0 +1,104 @@
+/**
+ * Content options for files
+ */
+export type ContentHeaders = {
+  contentType?: string
+  contentLanguage?: string
+  contentEncoding?: string
+  contentDisposition?: string
+  cacheControl?: string
+}
+
+/**
+ * Options for writing, moving and copying files
+ */
+export type WriteOptions = ContentHeaders & Record<string, any>
+
+/**
+ * Options for generating signed urls
+ */
+export type SignedUrlOptions = ContentHeaders & {
+  /**
+   * Signed URL expiration time in seconds
+   */
+  expiresIn?: number
+}
+
+/**
+ * Stats returned by the drive drivers
+ */
+export type DriveFileStats = {
+  size: number
+  modified: Date
+  isFile: boolean
+  etag?: string
+}
+
+/**
+ * Shape of the generic driver
+ */
+export interface DriverContract {
+  /**
+   * Name of the driver
+   */
+  name: string
+
+  /**
+   * A boolean to find if the location path exists or not
+   */
+  exists(location: string): Promise<boolean>
+
+  /**
+   * Returns the file contents as a buffer.
+   */
+  get(location: string): Promise<Buffer>
+
+  /**
+   * Returns the file contents as a stream
+   */
+  getStream(location: string): Promise<NodeJS.ReadableStream>
+
+  /**
+   * Returns the location path stats
+   */
+  getStats(location: string): Promise<DriveFileStats>
+
+  /**
+   * Returns a signed URL for a given location path
+   */
+  getSignedUrl(location: string, options?: SignedUrlOptions): Promise<string>
```

Comment:
Signed URLs can be for both download and upload on most cloud providers.
It could be less confusing to name this `getDownloadSignedUrl`.
So in the future an implementation of `getUploadSignedUrl` can be proposed.

Reply:
It also can be passed in the options `action: 'download' | 'upload'`

Reply:
Good idea! It should be pretty easy to (later) add an `action` property, to handle upload urls.
However, tests will have to be written first to avoid breaking the existing code.

I will add this to the TODO list of this package.